### PR TITLE
MINOR: migrate ReassignPartitionsIntegrationTest to use ClusterTestExtensions

### PR DESCRIPTION
**Pull request by FrankYang0529**
* Enable Kraft test for `ReassignPartitionsIntegrationTest`.
* Migrate `ReassignPartitionsIntegrationTest` to use `ClusterTestExtensions`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
@FrankYang0529 Could we use `ClusterTestExtensions` to enable `kraft`? It seems to me that using the new test framework is the better way. Otherwise, we may rewrite it again in the future.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 Could we use `ClusterTestExtensions` to enable `kraft`? It seems to me that using the new test framework is the better way. Otherwise, we may rewrite it again in the future.

Of course. I will dig into it tomorrow. Thanks for the suggestion.

**Comment by chia7712** [comment]
btw, it would be great if you can rewrite it by new test infra.

**Comment by FrankYang0529** [comment]
> btw, it would be great if you can rewrite it by new test infra.

Yeah, but this class need to assign different config to different broker. I will wait for https://github.com/apache/kafka/pull/15715.

https://github.com/apache/kafka/blob/025f9816f1a15d14aab25c9e8e5b03a87f0cefe2/tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsIntegrationTest.java#L754-L808

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+        waitForInterBrokerThrottle(asList(0, 1, 2, 3), interBrokerThrottle);
+
+        Map<TopicPartition, PartitionReassignmentState> finalAssignment = new HashMap<>();
+        finalAssignment.put(new TopicPartition("foo", 0),
+                new PartitionReassignmentState(asList(0, 3, 2), asList(0, 3, 2), true));
+        finalAssignment.put(new TopicPartition("baz", 2),
+                new PartitionReassignmentState(asList(3, 2, 1), asList(3, 2, 1), true));
+
+        // Wait for the assignment to complete
+        TestUtils.waitUntilTrue(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: We don't need to create `Admin` for each loop. Also, we ought to apply java version. For example:
```java
        try (Admin admin = Admin.create(Collections.singletonMap(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers()))) {
            org.apache.kafka.test.TestUtils.waitForCondition(() -> {
                Set<Integer> isr = admin.describeTopics(Collections.singleton(foo0.topic()))
                    .allTopicNames().get().get(foo0.topic()).partitions().stream()
                    .filter(p -> p.partition() == foo0.partition())
                    .flatMap(p -> p.isr().stream())
                    .map(Node::id).collect(Collectors.toSet());
                return isr.containsAll(Arrays.asList(0, 1, 2, 3));
            }, "Timed out while waiting for replica 3 to join the ISR");
        }
```

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+                    String configValue = Optional.ofNullable(brokerConfigs.get(throttleName)).map(ConfigEntry::value).orElse("-1");
+                    throttles.put(throttleName, Long.parseLong(configValue));
+                });
+                results.put(brokerId, throttles);
+            }
+        }
+        return results;
+    }
+
+    static class LogDirReassignment {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java:639]: It seems this is unused.

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+    ) {
+        System.out.println("==> cancelAssignment(adminClient, jsonString=" + jsonString);
+        try (Admin admin = Admin.create(Collections.singletonMap(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers()))) {
+            return cancelAssignment(admin, jsonString, preserveThrottles, 10000L, Time.SYSTEM);
+        } catch (ExecutionException | InterruptedException | JsonProcessingException | TerseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static class BrokerDirs {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: ditto.  unused

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+        String assignment = "{\"version\":1,\"partitions\":" +
+                "[{\"topic\":\"foo\",\"partition\":0,\"replicas\":[0,3,4],\"log_dirs\":[\"any\",\"any\",\"any\"]}" +
+                "]}";
+
+        // We will throttle replica 4 so that only replica 3 joins the ISR
+        setReplicationThrottleForPartitions(foo0);
+
+        // Execute the assignment and wait for replica 3 (only) to join the ISR
+        runExecuteAssignment(false, assignment, -1L, -1L);
+        TestUtils.waitUntilTrue(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: ditto. please consider using java version.

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,776 @@
...
+
+            BROKER_LEVEL_THROTTLES.forEach(throttle -> brokerConfig.put(throttle, -1L));
+
+            unthrottledBrokerConfigs.put(brokerId, brokerConfig);
+        });
+    }
+
+    private void verifyReplicaDeleted(TopicPartitionReplica topicPartitionReplica) throws ExecutionException, InterruptedException {
+        try (Admin admin = Admin.create(Collections.singletonMap(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers()))) {
+            TestUtils.waitUntilTrue(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: ditto

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+                    ? throttleConfig
+                    : unthrottledConfig;
+            throttledBrokerConfigs.put(brokerId, expectedThrottleConfig);
+        });
+        waitForBrokerLevelThrottles(throttledBrokerConfigs);
+    }
+
+    private void waitForBrokerLevelThrottles(Map<Integer, Map<String, Long>> targetThrottles) {
+        AtomicReference<Map<Integer, Map<String, Long>>> curThrottles = new AtomicReference<>(new HashMap<>());
+        TestUtils.waitUntilTrue(() -> {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: ditto

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+        } catch (ExecutionException | InterruptedException | JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void waitForVerifyAssignment(String jsonString,
+                                         Boolean preserveThrottles,
+                                         VerifyAssignmentResult expectedResult) {
+        final VerifyAssignmentResult[] latestResult = {null};
+        TestUtils.waitUntilTrue(
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: ditto

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsIntegrationTest.java [hunk 8]
```diff
@@ -64,40 +48,23 @@
...
-import static org.apache.kafka.tools.reassign.ReassignPartitionsCommand.cancelAssignment;
 import static org.apache.kafka.tools.reassign.ReassignPartitionsCommand.executeAssignment;
 import static org.apache.kafka.tools.reassign.ReassignPartitionsCommand.verifyAssignment;
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertTrue;
 
-@SuppressWarnings("ClassFanOutComplexity")
 @Timeout(300)
 public class ReassignPartitionsIntegrationTest extends QuorumTestHarness {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsIntegrationTest.java]: `unthrottledBrokerConfigs` is unused in this test now. could you please remove it?

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsIntegrationTest.java [hunk 9]
```diff
@@ -119,93 +86,6 @@ public void tearDown() {
...
-            describeBrokerLevelThrottles(unthrottledBrokerConfigs.keySet()));
-
-        // Verify that partitions are removed from brokers no longer assigned
-        verifyReplicaDeleted(foo0, 2);
-        verifyReplicaDeleted(bar0, 1);
-    }
-
     @ParameterizedTest
     @ValueSource(strings = {"zk", "kraft"})
     public void testHighWaterMarkAfterPartitionReassignment(String quorum) throws Exception {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsIntegrationTest.java]: It seems the check list of this case is the new leader replica has synced the HWM. Hence, we can rewrite it by following steps.

1. create a partition with 2 replicas (for example: 0,1)
2. add some records
3. create new reassignment with a new replica leader (for example: 2,0,1)
4. wait for the new leader to be on board
5. check the HWD (by `Admin#listOffsets`)

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 nice patch!

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for the review and suggestion. I addressed all comments. Thank you.

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,820 @@
...
+        consumerProps.put(ConsumerConfig.GROUP_ID_CONFIG, "group");
+        consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProps.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
+        TopicPartition part = new TopicPartition("baz", 2);
+        try (Consumer<byte[], byte[]> consumer = new KafkaConsumer<>(consumerProps, new ByteArrayDeserializer(), new ByteArrayDeserializer())) {
+            consumer.assign(singleton(part));
+            List<ConsumerRecord<byte[], byte[]>> allRecords = new ArrayList<>();
+            TestUtils.waitForCondition(() -> {
+                ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(100L));
+                System.out.println("records.count " + records.count());
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: Could we remove those debug messages?

Comment (chia7712) [reply tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: It seems there are many similar `System.out.println` in this test

Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: Removed all `System.out.println`. Thank you.

**Comment by FrankYang0529** [comment]
The CI is finished and failed test cases are not related to this PR.

OffsetsApiIntegrationTest.testAlterSinkConnectorOffsets: https://issues.apache.org/jira/browse/KAFKA-15892
OffsetsApiIntegrationTest.testAlterSinkConnectorOffsetsZombieSinkTasks: https://issues.apache.org/jira/browse/KAFKA-15917
TransactionsTest.testFencingOnCommit: https://issues.apache.org/jira/browse/KAFKA-12567
ExactlyOnceSourceIntegrationTest.testSeparateOffsetsTopic: https://issues.apache.org/jira/browse/KAFKA-14089
ConsumerBounceTest.testConsumptionWithBrokerFailures: https://issues.apache.org/jira/browse/KAFKA-15146
ReplicationQuotasTest.shouldThrottleOldSegments: https://issues.apache.org/jira/browse/KAFKA-16635
DelegationTokenEndToEndAuthorizationWithOwnerTest: https://issues.apache.org/jira/browse/KAFKA-15411

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,811 @@
...
+                "Latest throttles were " + curThrottles.get());
+    }
+
+    /**
+     * Describe the broker-level throttles in the cluster.
+     *
+     * @return                A map whose keys are broker IDs and whose values are throttle
+     *                        information.  The nested maps are keyed on throttle name.
+     */
+    private Map<Integer, Map<String, Long>> describeBrokerLevelThrottles(Admin admin, Collection<Integer> brokerIds) throws ExecutionException, InterruptedException {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: Could we simplify this function by following style?
```java
    private Map<Integer, Map<String, Long>> describeBrokerLevelThrottles(Admin admin, Collection<Integer> brokerIds) {
        return brokerIds.stream().collect(Collectors.toMap(Function.identity(), brokerId -> {
            ConfigResource brokerResource = new ConfigResource(ConfigResource.Type.BROKER, brokerId.toString());
            Config brokerConfigs = assertDoesNotThrow(() -> admin.describeConfigs(singleton(brokerResource)).values()
                .get(brokerResource)
                .get());
            return BROKER_LEVEL_THROTTLES.stream().collect(Collectors.toMap(Function.identity(),
                name -> Optional.ofNullable(brokerConfigs.get(name)).map(e -> Long.parseLong(e.value())).orElse(-1L)));
        }));
    }
```

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,811 @@
...
+            unthrottledBrokerConfigs.put(brokerId, brokerConfig);
+        });
+    }
+
+    private void verifyReplicaDeleted(TopicPartitionReplica topicPartitionReplica) throws InterruptedException {
+        try (Admin admin = Admin.create(Collections.singletonMap(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers()))) {
+            TestUtils.waitForCondition(
+                    () -> {
+                        try {
+                            TopicDescription topicDescription = admin.describeTopics(singleton(topicPartitionReplica.topic())).topicNameValues().get(topicPartitionReplica.topic()).get();
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: We can use `assertDoesNotThrow` to convert the checked exception to unchecked version.

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,811 @@
...
+import static org.apache.kafka.tools.reassign.ReassignPartitionsCommand.verifyAssignment;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@ClusterTestDefaults(brokers = 5, disksPerBroker = 3, serverProperties = {
+        @ClusterConfigProperty(key = REPLICA_FETCH_BACKOFF_MS_CONFIG, value = "100"),
+        @ClusterConfigProperty(key = AUTO_LEADER_REBALANCE_ENABLE_CONFIG, value = "false"),
+        @ClusterConfigProperty(key = REPLICA_LAG_TIME_MAX_MS_CONFIG, value = "1000"),
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java:105]: not sure which test cases require this config. Could you point out it for me?

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,811 @@
...
+        // Now cancel the assignment and verify that the partition is removed from cancelled replicas
+        assertEquals(new AbstractMap.SimpleImmutableEntry<>(singleton(foo0), Collections.emptySet()), runCancelAssignment(assignment, true));
+        verifyReplicaDeleted(new TopicPartitionReplica(foo0.topic(), foo0.partition(), 3));
+        verifyReplicaDeleted(new TopicPartitionReplica(foo0.topic(), foo0.partition(), 4));
+    }
+
+    /**
+     * Test moving partitions between directories.
+     */
+    @ClusterTest(types = {Type.ZK, Type.KRAFT}) // JBOD not yet implemented for KRaft
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: `BOD not yet implemented for KRaft` is this comment a bit awkward ?

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,811 @@
...
+        properties.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers());
+        try (KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(properties, new ByteArraySerializer(), new ByteArraySerializer())) {
+            IntStream.range(0, numMessages).forEach(i -> {
+                ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(topic, partition, null, new byte[10000]);
+                assertDoesNotThrow(() -> producer.send(record).get());
+            });
+        }
+    }
+
+    private void executeAndVerifyReassignment() throws ExecutionException, InterruptedException {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: `ExecutionException` is unused

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 16]
```diff
@@ -0,0 +1,811 @@
...
+            assertEquals(unthrottledBrokerConfigs,
+                    describeBrokerLevelThrottles(admin, unthrottledBrokerConfigs.keySet()));
+        }
+
+        // Verify that partitions are removed from brokers no longer assigned
+        verifyReplicaDeleted(new TopicPartitionReplica(foo0.topic(), foo0.partition(), 2));
+        verifyReplicaDeleted(new TopicPartitionReplica(bar0.topic(), bar0.partition(), 1));
+    }
+
+    private final Map<Integer, Map<String, Long>> unthrottledBrokerConfigs = new HashMap<>(); {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: this code style is uncommon. Could we rewrite it by lambda and move it up to line#113?

File: tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java [hunk 17]
```diff
@@ -0,0 +1,811 @@
...
+            Boolean preserveThrottles
+    ) {
+        try (Admin admin = Admin.create(Collections.singletonMap(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, clusterInstance.bootstrapServers()))) {
+            return cancelAssignment(admin, jsonString, preserveThrottles, 10000L, Time.SYSTEM);
+        } catch (ExecutionException | InterruptedException | JsonProcessingException | TerseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static class BrokerDirs {
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommandTest.java]: those static class can be private

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR!

**Comment by FrankYang0529** [comment]
@chia7712 I address all comments. Thanks for the review.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**